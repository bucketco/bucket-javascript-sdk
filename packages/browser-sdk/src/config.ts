import { version } from "../package.json";

export const API_BASE_URL = "https://front.bucket.co";
export const APP_BASE_URL = "https://app.bucket.co";
export const SSE_REALTIME_BASE_URL = "https://livemessaging.bucket.co";

export const SDK_VERSION_HEADER_NAME = "reflag-sdk-version";

export const SDK_VERSION = `browser-sdk/${version}`;
export const FEATURE_EVENTS_PER_MIN = 1;
