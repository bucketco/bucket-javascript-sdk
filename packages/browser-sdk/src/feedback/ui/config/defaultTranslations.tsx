import { FeedbackTranslations } from "../types";
/**
 * {@includeCode ./defaultTranslations.tsx}
 */
export const DEFAULT_TRANSLATIONS: FeedbackTranslations = {
  DefaultQuestionLabel: "How satisfied are you with this feature?",
  QuestionPlaceholder: "Write a comment",
  ScoreStatusDescription: "Pick a score and leave a comment",
  ScoreStatusLoading: "Saving score, please wait...",
  ScoreStatusReceived: "Score has been received!",
  ScoreVeryDissatisfiedLabel: "Very dissatisfied (1/5)",
  ScoreDissatisfiedLabel: "Dissatisfied (2/5)",
  ScoreNeutralLabel: "Neutral (3/5)",
  ScoreSatisfiedLabel: "Satisfied (4/5)",
  ScoreVerySatisfiedLabel: "Very satisfied (5/5)",
  SuccessMessage: "Feedback received, thank you!",
  SendButton: "Send feedback",
};
