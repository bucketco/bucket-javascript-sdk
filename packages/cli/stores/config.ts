import { Ajv, ValidateFunction } from "ajv";
import { findUp } from "find-up";
import JSON5 from "json5";
import { readFile, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import {
  CONFIG_FILE_NAME,
  DEFAULT_API_URL,
  DEFAULT_BASE_URL,
  DEFAULT_TYPES_OUTPUT,
  SCHEMA_URL,
} from "../utils/constants.js";
import { ConfigValidationError, handleError } from "../utils/errors.js";
import { stripTrailingSlash } from "../utils/path.js";

export const keyFormats = [
  "custom",
  "pascalCase",
  "camelCase",
  "snakeCaseUpper",
  "snakeCaseLower",
  "kebabCaseUpper",
  "kebabCaseLower",
] as const;

export type KeyFormat = (typeof keyFormats)[number];

type Config = {
  $schema: string;
  baseUrl: string;
  apiUrl: string;
  appId: string | undefined;
  typesOutput: string;
  keyFormat: KeyFormat;
};

const defaultConfig: Config = {
  $schema: SCHEMA_URL,
  baseUrl: DEFAULT_BASE_URL,
  apiUrl: DEFAULT_API_URL,
  appId: undefined,
  typesOutput: DEFAULT_TYPES_OUTPUT,
  keyFormat: "custom",
};

class ConfigStore {
  protected config: Config = { ...defaultConfig };
  protected configPath: string | undefined;
  protected projectPath: string | undefined;
  protected validateConfig: ValidateFunction | undefined;

  async initialize() {
    await this.createValidator();
    await this.loadConfigFile();
  }

  protected async createValidator() {
    try {
      // Using current config store file, resolve the schema.json path
      const filePath = fileURLToPath(import.meta.url);
      const schemaPath = join(
        filePath.substring(0, filePath.indexOf("cli") + 3),
        "schema.json",
      );
      const content = await readFile(schemaPath, "utf-8");
      const parsed = JSON5.parse<Config>(content);
      const ajv = new Ajv();
      this.validateConfig = ajv.compile(parsed);
    } catch {
      void handleError(new Error("Failed to load the config schema"), "Config");
    }
  }

  protected async loadConfigFile() {
    if (!this.validateConfig) {
      void handleError(new Error("Failed to load the config schema"), "Config");
      return;
    }

    try {
      const packageJSONPath = await findUp("package.json");
      this.configPath = await findUp(CONFIG_FILE_NAME);
      this.projectPath = dirname(
        this.configPath ?? packageJSONPath ?? process.cwd(),
      );

      if (!this.configPath) return;

      const content = await readFile(this.configPath, "utf-8");
      const parsed = JSON5.parse<Partial<Config>>(content);
      parsed.baseUrl = stripTrailingSlash(parsed.baseUrl?.trim());
      parsed.apiUrl = stripTrailingSlash(parsed.apiUrl?.trim());

      if (!this.validateConfig!(parsed)) {
        void handleError(
          new ConfigValidationError(this.validateConfig!.errors),
          "Config",
        );
      }

      this.config = { ...this.config, ...parsed };
    } catch {
      // No config file found
    }
  }

  /**
   * Create a new config file with initial values.
   * @param overwrite If true, overwrites existing config file. Defaults to false
   */
  async saveConfigFile(overwrite = false) {
    const configWithoutDefaults: Partial<Config> = {
      ...this.config,
    };

    // Only include non-default values and $schema
    for (const untypedKey in configWithoutDefaults) {
      const key = untypedKey as keyof Config;
      if (
        !["$schema"].includes(key) &&
        configWithoutDefaults[key] === defaultConfig[key]
      ) {
        delete configWithoutDefaults[key];
      }
    }

    const configJSON = JSON.stringify(configWithoutDefaults, null, 2);

    if (this.configPath && !overwrite) {
      throw new Error("Config file already exists");
    }

    if (this.configPath) {
      await writeFile(this.configPath, configJSON);
    } else {
      // Write to the project path
      const packageJSONPath = await findUp("package.json");
      this.projectPath = dirname(packageJSONPath ?? process.cwd());
      this.configPath = join(this.projectPath, CONFIG_FILE_NAME);
      await writeFile(this.configPath, configJSON);
    }
  }

  getConfig(): Config;
  getConfig<K extends keyof Config>(key: K): Config[K];
  getConfig<K extends keyof Config>(key?: K) {
    return key ? this.config?.[key] : this.config;
  }

  getConfigPath() {
    return this.configPath;
  }

  getProjectPath() {
    return this.projectPath ?? process.cwd();
  }

  setConfig(newConfig: Partial<Config>) {
    // Update the config with new values skipping undefined values
    for (const untypedKey in newConfig) {
      const key = untypedKey as keyof Config;
      if (newConfig[key] === undefined) continue;
      (this.config as any)[key] = newConfig[key];
    }
  }
}

export const configStore = new ConfigStore();
