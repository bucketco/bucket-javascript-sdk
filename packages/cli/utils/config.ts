import { readJson, writeJson } from "fs-extra/esm";

import { CONFIG_FILE } from "./constants.js";

type Config = {
  sessionCookies?: string;
  appId?: string;
  envId?: string;
};

let config: Config = {};

/**
 * Instantly return a specified key's value or the entire config object.
 */
export function getConfig(): Config;
export function getConfig(key: keyof Config): string | undefined;
export function getConfig(key?: keyof Config) {
  return key ? config[key] : config;
}

/**
 * Read the config file and return either a specified key's value or the entire config object.
 */
export async function readConfigFile(): Promise<Config>;
export async function readConfigFile(
  key: keyof Config,
): Promise<string | undefined>;
export async function readConfigFile(key?: keyof Config) {
  try {
    config = await readJson(CONFIG_FILE);
    return key ? config[key] : config;
  } catch (error) {
    return {};
  }
}

/**
 * Write a new value to the config file.
 */
export async function writeConfigFile(
  key: keyof Config,
  value: string | undefined,
) {
  const config = await readConfigFile();
  config[key] = value;
  await writeJson(CONFIG_FILE, config);
}
