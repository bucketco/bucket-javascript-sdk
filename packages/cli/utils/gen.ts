export const genDTS = (keys: string[]) => {
  return /* ts */ `
// DO NOT EDIT THIS FILE. IT IS GENERATED BY THE BUCKET-CLI AND WILL BE OVERWRITTEN.
const availableFeatures = [
  "${keys.join('",\n  "')}"
] as const;

type ArrayToRecord<T extends readonly string[]> = {
  [Key in T[number]]: boolean;
};

export type AvailableFeatures = ArrayToRecord<typeof availableFeatures>;

declare module "@bucketco/react-sdk" {
  interface Features extends AvailableFeatures {}
}
`.trim();
};
