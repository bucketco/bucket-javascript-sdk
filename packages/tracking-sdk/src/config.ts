import { version } from "../package.json";

export const API_HOST = "https://tracking.bucket.co";
export const SSE_REALTIME_HOST = "https://livemessaging.bucket.co";

export const SDK_VERSION_HEADER_NAME = "bucket-sdk-version";

export const SDK_VERSION = version;
export const FLAG_EVENTS_PER_MIN = 1;
