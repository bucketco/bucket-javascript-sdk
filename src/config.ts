export const TRACKING_HOST = "http://localhost:3200";
