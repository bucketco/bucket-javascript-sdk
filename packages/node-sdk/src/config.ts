import { readFileSync } from "fs";

import { version } from "../package.json";

import { Logger } from "./types";
import { ok } from "./utils";

export const API_HOST = "https://front.bucket.co";
export const SDK_VERSION_HEADER_NAME = "bucket-sdk-version";
export const SDK_VERSION = `node-sdk/${version}`;
export const API_TIMEOUT_MS = 5000;

export const BUCKET_LOG_PREFIX = "[Bucket]";

export const FEATURE_EVENTS_PER_MIN = 1;
export const FEATURES_REFETCH_MS = 60 * 1000; // re-fetch every 60 seconds

export const BATCH_MAX_SIZE = 100;
export const BATCH_INTERVAL_MS = 60 * 1000;

function parseOverrides(config: object | undefined) {
  if (!config) return {};
  if (
    "featureOverrides" in config &&
    typeof config.featureOverrides === "object"
  ) {
    const overrides = config.featureOverrides as object;
    Object.entries(overrides).forEach(([key, value]) => {
      ok(
        typeof value === "boolean",
        `invalid type "${typeof value}" for key ${key}, expected boolean`,
      );
    });
    return overrides;
  }
  return {};
}

function loadConfigFile(file: string) {
  const configJson = readFileSync(file, "utf-8");
  const config = JSON.parse(configJson);

  ok(typeof config === "object", "config must be an object");
  const { secretKey, logLevel, offline, host } = config;
  console.log("secretKey", secretKey);
  ok(
    typeof secretKey === "undefined" || typeof secretKey === "string",
    "secret must be a string",
  );
  ok(
    typeof logLevel === "undefined" ||
      (typeof logLevel === "string" && LOG_LEVELS.includes(logLevel)),
    `logLevel must one of ${LOG_LEVELS.join(", ")}`,
  );
  ok(
    typeof offline === "undefined" || typeof offline === "boolean",
    "offline must be a boolean",
  );

  return {
    featureOverrides: parseOverrides(config),
    secretKey,
    logLevel,
    offline,
    host,
  };
}

function loadEnvVars() {
  const secretKey = process.env.BUCKET_SECRET_KEY;
  const enabledFeatures = process.env.BUCKET_FEATURES_ENABLED;
  const disabledFeatures = process.env.BUCKET_FEATURES_DISABLED;
  const logLevel = process.env.BUCKET_LOG_LEVEL;
  const host = process.env.BUCKET_HOST;
  const offline =
    process.env.BUCKET_OFFLINE !== undefined
      ? ["true", "on"].includes(process.env.BUCKET_OFFLINE)
      : undefined;

  let featureOverrides: Record<string, boolean> = {};
  if (enabledFeatures) {
    featureOverrides = enabledFeatures.split(",").reduce(
      (acc, f) => {
        const key = f.trim();
        if (key) acc[key] = true;
        return acc;
      },
      {} as Record<string, boolean>,
    );
  }

  if (disabledFeatures) {
    featureOverrides = {
      ...featureOverrides,
      ...disabledFeatures.split(",").reduce(
        (acc, f) => {
          const key = f.trim();
          if (key) acc[key] = false;
          return acc;
        },
        {} as Record<string, boolean>,
      ),
    };
  }

  return { secretKey, featureOverrides, logLevel, offline, host };
}

export function loadConfig(file?: string) {
  let fileConfig;
  if (file) {
    fileConfig = loadConfigFile(file);
  }

  const envConfig = loadEnvVars();

  return {
    secretKey: envConfig.secretKey || fileConfig?.secretKey,
    logLevel: envConfig.logLevel || fileConfig?.logLevel,
    offline: envConfig.offline ?? fileConfig?.offline,
    host: envConfig.host ?? fileConfig?.host,
    featureOverrides: {
      ...fileConfig?.featureOverrides,
      ...envConfig.featureOverrides,
    },
  };
}

const LOG_LEVELS = ["DEBUG", "INFO", "WARN", "ERROR"];

export function applyLogLevel(logger: Logger, logLevel?: string) {
  switch (logLevel) {
    case "debug":
      return {
        debug: logger.debug,
        info: logger.info,
        warn: logger.warn,
        error: logger.error,
      };
    case "info":
      return {
        debug: () => void 0,
        info: logger.info,
        warn: logger.warn,
        error: logger.error,
      };
    case "warn":
      return {
        debug: () => void 0,
        info: () => void 0,
        warn: logger.warn,
        error: logger.error,
      };
    case "error":
      return {
        debug: () => void 0,
        info: () => void 0,
        warn: () => void 0,
        error: logger.error,
      };
    default:
      return {
        debug: () => void 0,
        info: () => void 0,
        warn: () => void 0,
        error: () => void 0,
      };
  }
}
