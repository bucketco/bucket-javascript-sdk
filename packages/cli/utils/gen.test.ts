import { describe, expect, it } from "vitest";

import { FeatureDef } from "./config.js";
import { genDTS, genJs } from "./gen.js";

describe("genFeatureTypes", () => {
  const features: FeatureDef[] = [
    { key: "feat-1", access: true, configType: undefined },
    { key: "FEAT URE2", access: false, configType: "string" },
    {
      key: "feature3",
      access: true,
      configType: {
        aiModel: "string",
        prompt: "string",
      },
      fallback: {
        isEnabled: true,
        config: {
          type: "number", // this will trip up a simple output writer
          aiModel: "gtp3.5-turbo",
          prompt: "Explain in conversational language",
        },
      },
    },
  ];

  it("should generate correct TypeScript declaration", () => {
    const output = genDTS(features);
    expect(output).toMatchSnapshot();
  });

  it("should generate correct javascript file", () => {
    const output = genJs(features);
    expect(output).toMatchSnapshot();
  });
});
