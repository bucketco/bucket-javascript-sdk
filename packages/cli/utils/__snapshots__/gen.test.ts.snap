// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`genFeatureTypes > should generate correct TypeScript output for browser 1`] = `
"// DO NOT EDIT THIS FILE. IT IS GENERATED BY THE BUCKET CLI AND WILL BE OVERWRITTEN.

export const generatedFeatures: string[];
export interface GeneratedFeatureTypes {
  "feat-1": {
    key: "feat-1", 
    access: true, 
    config: undefined
  }, 
  "FEAT URE2": {
    key: "FEAT URE2", 
    access: false, 
    config: string
  }, 
  feature3: {
    key: "feature3", 
    access: true, 
    config: {
      aiModel: string, 
      prompt: string
    }
  }
}"
`;

exports[`genFeatureTypes > should generate correct TypeScript output for react 1`] = `
"// DO NOT EDIT THIS FILE. IT IS GENERATED BY THE BUCKET CLI AND WILL BE OVERWRITTEN.

export const generatedFeatures: string[];
export interface GeneratedFeatureTypes {
  "feat-1": {
    key: "feat-1", 
    access: true, 
    config: undefined
  }, 
  "FEAT URE2": {
    key: "FEAT URE2", 
    access: false, 
    config: string
  }, 
  feature3: {
    key: "feature3", 
    access: true, 
    config: {
      aiModel: string, 
      prompt: string
    }
  }
}"
`;
